### PR TITLE
Split static

### PR DESCRIPTION
**Pull request by ocefpaf**
Attempt to solve https://github.com/conda-forge/openblas-feedstock/issues/69

This did not work locally and I'm not sure what is wrong (Writing split-recipes is a new thing to me :smile:). @msarahan and @isuruf I appreciate any advice you may have to make this work.

It fails with a `KeyError: 'lib/libopenblasp-r0.3.6.so'` when finalizing the package. Full error:

```
ERROR :: get_rpaths_raw()=['/opt/conda/conda-bld/openblas_1559093006932/_build_env/lib:/opt/conda/conda-bld/openblas_1559093006932/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib'] and patchelf=/opt/conda/conda-bld/openblas_1559093006932/_build_env/lib:/opt/conda/conda-bld/openblas_1559093006932/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib disagree for /opt/conda/conda-bld/openblas_1559093006932/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libopenblasp-r0.3.6.so :: 
Warning: rpath /opt/conda/conda-bld/openblas_1559093006932/_build_env/lib is outside prefix /opt/conda/conda-bld/openblas_1559093006932/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac (removing it)
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 445, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 436, in execute
    verify=args.verify, variants=args.variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2299, in build_tree
    notest=notest,
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1596, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 958, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 823, in post_process_files
    post_build(m, new_files, build_python=python)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/post.py", line 1070, in post_build
    check_overlinking(m, files)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/post.py", line 1009, in check_overlinking
    m.config.channel_urls)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/post.py", line 930, in check_overlinking_impl
    needed = needed_dsos_for_file[f]
KeyError: 'lib/libopenblasp-r0.3.6.so'
```

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -21,26 +20,46 @@ requirements:
...
+        - python -c "import ctypes; ctypes.cdll['${PREFIX}/lib/libopenblas{{ SHLIB_EXT }}']"  # [not win]
+        - if not exist %PREFIX%\\Library\\lib\\openblas.lib exit 1  # [win]
+        - if not exist %PREFIX%\\Library\\bin\\openblas.dll exit 1  # [win]
+
+  - name: openblas
+    files:
+      - include/*.h
+      - lib/*.a
+      - lib/pkgconfig/*.pc
+      - lib/cmake/openblas/*
```
Comment (ocefpaf) [inline recipe/meta.yaml]: Not sure what to do on Windows. Would a `/Library/*` work for all sub-directories?
Do we need a `libopenblas` on Windows? If so it should be the `.dll`, the `.lib` or both?

Comment (isuruf) [reply recipe/meta.yaml]: Just the `.dll`

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -21,26 +20,56 @@ requirements:
...
+    build:
+      skip: true  # [win and vc<14]
+      run_exports:
+        - {{ pin_subpackage("libopenblas") }}
+    requirements:
+      build:
+        - {{ compiler("c") }}
+        - {{ compiler("fortran") }}
+    files:
+      - lib/*{{ SHLIB_EXT }}*  # [not win]
```
Comment (isuruf) [inline recipe/meta.yaml]: We don't want `lib/libopenblas${SHLIB_EXT}` in `libopenblas`

File: recipe/meta.yaml [hunk 3]
```diff
@@ -5,11 +5,10 @@ package:
   version: {{ version }}
 
 source:
-  fn: openblas-{{ version }}.tar.gz
   url: https://github.com/xianyi/OpenBLAS/archive/v{{ version }}.tar.gz
   sha256: e64c8fe083832ffbc1459ab6c72f71d53afd3b36e8497c922a15a06b72e9002f
   patches:
-    - 0002-Fix-gfortran-detection-for-ctng-based-cross-compilers.patch    # [c_compiler == 'gcc']
```
Comment (isuruf) [inline recipe/meta.yaml]: Can you revert this change?

Comment (ocefpaf) [reply recipe/meta.yaml]: Sure. I was just testing locally and should've not pushed it.

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: I probably misunderstood then. I though that all `.so`/`.dylib` should be in `libopenblas`. Why not?

File: recipe/meta.yaml [hunk 4]
```diff
@@ -21,26 +20,56 @@ requirements:
...
+outputs:
+  - name: libopenblas
+    build:
+      skip: true  # [win and vc<14]
+      run_exports:
+        - {{ pin_subpackage("libopenblas") }}
+    requirements:
+      build:
+        - {{ compiler("c") }}
+        - {{ compiler("fortran") }}
```
Comment (ocefpaf) [inline recipe/meta.yaml]: It is unclear to me, coming from a RPM world where we build and then list the files in each sub-package, why we need two build sections? But that seems to work :smile:

**Comment by ocefpaf** [comment]
This is what we have in `libopenblas`:

```
.
├── [4.0K]  info
│   ├── [3.3K]  about.json
│   ├── [ 608]  files
│   ├── [   0]  git
│   ├── [ 123]  hash_input.json
│   ├── [ 290]  has_prefix
│   ├── [ 298]  index.json
│   ├── [1.5K]  LICENSE.txt
│   ├── [6.5K]  paths.json
│   ├── [4.0K]  recipe
│   │   ├── [ 696]  conda_build_config.yaml
│   │   ├── [1.5K]  meta.yaml
│   │   └── [4.0K]  parent
│   │       ├── [ 308]  0002-Fix-gfortran-detection-for-ctng-based-cross-compilers.patch
│   │       ├── [ 666]  0003-Fix-power-prefetch.patch
│   │       ├── [ 567]  bld.bat
│   │       ├── [2.1K]  build.sh
│   │       ├── [1.7K]  gh2111.patch
│   │       └── [2.0K]  meta.yaml
│   ├── [  40]  run_exports.json
│   └── [4.0K]  test
│       └── [ 128]  run_test.sh
└── [4.0K]  lib
    ├── [  16]  libasan.so -> libasan.so.5.0.0
    ├── [  16]  libasan.so.5 -> libasan.so.5.0.0
    ├── [ 13M]  libasan.so.5.0.0
    ├── [  18]  libatomic.so -> libatomic.so.1.2.0
    ├── [  18]  libatomic.so.1 -> libatomic.so.1.2.0
    ├── [180K]  libatomic.so.1.2.0
    ├── [ 132]  libgcc_s.so
    ├── [801K]  libgcc_s.so.1
    ├── [  20]  libgfortran.so -> libgfortran.so.4.0.0
    ├── [  20]  libgfortran.so.4 -> libgfortran.so.4.0.0
    ├── [5.7M]  libgfortran.so.4.0.0
    ├── [  16]  libgomp.so -> libgomp.so.1.0.0
    ├── [  16]  libgomp.so.1 -> libgomp.so.1.0.0
    ├── [916K]  libgomp.so.1.0.0
    ├── [  15]  libitm.so -> libitm.so.1.0.0
    ├── [  15]  libitm.so.1 -> libitm.so.1.0.0
    ├── [941K]  libitm.so.1.0.0
    ├── [  16]  liblsan.so -> liblsan.so.0.0.0
    ├── [  16]  liblsan.so.0 -> liblsan.so.0.0.0
    ├── [ 13M]  liblsan.so.0.0.0
    ├── [ 29M]  libopenblasp-r0.3.6.so
    ├── [  22]  libopenblas.so -> libopenblasp-r0.3.6.so
    ├── [  22]  libopenblas.so.0 -> libopenblasp-r0.3.6.so
    ├── [  20]  libquadmath.so -> libquadmath.so.0.0.0
    ├── [  20]  libquadmath.so.0 -> libquadmath.so.0.0.0
    ├── [845K]  libquadmath.so.0.0.0
    ├── [  16]  libtsan.so -> libtsan.so.0.0.0
    ├── [  16]  libtsan.so.0 -> libtsan.so.0.0.0
    ├── [ 15M]  libtsan.so.0.0.0
    ├── [  17]  libubsan.so -> libubsan.so.1.0.0
    ├── [  17]  libubsan.so.1 -> libubsan.so.1.0.0
    └── [ 11M]  libubsan.so.1.0.0
```

And this is the contents in `openblas` (that depends on `libopenblas`):

```
.
├── [4.0K]  include
│   ├── [ 46K]  cblas.h
│   ├── [ 47K]  f77blas.h
│   ├── [4.3K]  lapacke_config.h
│   ├── [1.2M]  lapacke.h
│   ├── [ 474]  lapacke_mangling.h
│   ├── [ 32K]  lapacke_utils.h
│   └── [4.9K]  openblas_config.h
├── [4.0K]  info
│   ├── [3.2K]  about.json
│   ├── [ 311]  files
│   ├── [   0]  git
│   ├── [ 123]  hash_input.json
│   ├── [ 142]  has_prefix
│   ├── [ 331]  index.json
│   ├── [1.5K]  LICENSE.txt
│   ├── [2.6K]  paths.json
│   └── [4.0K]  recipe
│       ├── [ 308]  0002-Fix-gfortran-detection-for-ctng-based-cross-compilers.patch
│       ├── [ 666]  0003-Fix-power-prefetch.patch
│       ├── [ 567]  bld.bat
│       ├── [2.1K]  build.sh
│       ├── [ 696]  conda_build_config.yaml
│       ├── [1.7K]  gh2111.patch
│       ├── [1.6K]  meta.yaml
│       └── [2.0K]  meta.yaml.template
└── [4.0K]  lib
    ├── [4.0K]  cmake
    │   └── [4.0K]  openblas
    │       ├── [ 174]  OpenBLASConfig.cmake
    │       └── [ 284]  OpenBLASConfigVersion.cmake
    ├── [  21]  libopenblas.a -> libopenblasp-r0.3.6.a
    ├── [ 50M]  libopenblasp-r0.3.6.a
    └── [4.0K]  pkgconfig
        └── [ 552]  openblas.pc
```

@isuruf it is unclear to me your https://github.com/conda-forge/openblas-feedstock/pull/70#discussion_r288424931, can you elaborate?

@jcrist the sizes of the packages are:

```
  17M May 29 07:05 libopenblas-0.3.6-h6e990d7_2.tar.bz2
 8.2M May 29 07:06 openblas-0.3.6-h6e990d7_2.tar.bz2
```

File: recipe/meta.yaml [hunk 5]
```diff
@@ -52,6 +54,7 @@ outputs:
     files:
       - include/*.h  # [not win]
       - lib/*.a  # [not win]
+      - lib/libopenblas${SHLIB_EXT}  # [not win]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: I'm still confused @isuruf `libopenblas${SHLIB_EXT}` is a symlink to `libopenblasp-r0.3.6${SHLIB_EXT}`, right? Why they have to be in separate packages? Also, size wise it does not make any difference. I guess that I'm worried that packages searching for `libopenblas${SHLIB_EXT}` will be forced to install the full package instead of just `libopenblas` b/c of a symlink.

Comment (isuruf) [reply recipe/meta.yaml]: > libopenblas${SHLIB_EXT} is a symlink to libopenblasp-r0.3.6${SHLIB_EXT}, right? 

Yes

> I guess that I'm worried that packages searching for libopenblas${SHLIB_EXT} will be forced to install the full package instead of just libopenblas b/c of a symlink.

No package will be looking for `libopenblas.so`. They will look for `libopenblas.so.0` which is the SONAME of the library.

Reason for `libopenblas.so` not being in `libopenblas` package is to avoid accidental linking to `libopenblas.so` by other packages. This is what debian and other package managers do. Compare https://packages.debian.org/jessie/amd64/libopenblas-base/filelist and https://packages.debian.org/jessie/amd64/libopenblas-dev/filelist

Comment (ocefpaf) [reply recipe/meta.yaml]: Following that logic should all the symlinks be in `openblas` instead?

Comment (isuruf) [reply recipe/meta.yaml]: No. We need the concrete file `libopenblasp-r0.3.6.so` and the SONAME library `libopenblas.so.0` in `libopenblas` because downstream packages will be looking for the SONAME library `libopenblas.so.0` even though it is a symlink.

Comment (ocefpaf) [reply recipe/meta.yaml]: Also, this diverges from what is happening in `defaults`: https://github.com/AnacondaRecipes/openblas-feedstock/blob/master/recipe/meta.yaml#L44

I'm OK with that change but we need to be aware of the package differences, specially b/c we are not building with `strict` and some problems may appear if `defaults` `libopenblas` is selected in our builds! (An important case for us to start using `strict` in our builds ASAP.)

Comment (isuruf) [reply recipe/meta.yaml]: Even if `defaults` `libopenblas` is used at runtime, that's fine because they have all the files we have in `libopenblas` and more files. It would be problematic if they had less files.

Comment (ocefpaf) [reply recipe/meta.yaml]: I guess that when you said:

    Reason for libopenblas.so not being in libopenblas package is to avoid accidental
    linking to libopenblas.so by other packages.

I thought that if a package built with `defaults` accidentally links to `libopenblas.so` would it be broken if we try to run with out `libopenblas`?

(I'm not against the change but worried if we should start using `stritc` in the builds or not.)

Comment (isuruf) [reply recipe/meta.yaml]: > I thought that if a package built with defaults accidentally links to libopenblas.so would it be broken if we try to run with out libopenblas?

Even if they link accidentally link to `libopenblas.so`, the linked library or executable will look for `libopenblas.so.0` at load time because that is the SONAME of `libopenblas.so`.

File: recipe/meta.yaml [hunk 6]
```diff
@@ -21,26 +20,59 @@ requirements:
...
+      skip: true  # [win and vc<14]
+    files:
+      - include/*.h  # [not win]
+      - lib/*.a  # [not win]
+      - lib/libopenblas${SHLIB_EXT}  # [not win]
+      - lib/pkgconfig/*.pc  # [not win]
+      - lib/cmake/openblas/*  # [not win]
+      - Library/include/*  # [win]
+      - Library/lib/*  # [win]
+      - Library/share/*  # [win]
```
Comment (isuruf) [inline recipe/meta.yaml]: Is this needed?

File: recipe/meta.yaml [hunk 7]
```diff
@@ -21,26 +20,59 @@ requirements:
...
+    build:
+      skip: true  # [win and vc<14]
+    files:
+      - include/*.h  # [not win]
+      - lib/*.a  # [not win]
+      - lib/libopenblas${SHLIB_EXT}  # [not win]
+      - lib/pkgconfig/*.pc  # [not win]
+      - lib/cmake/openblas/*  # [not win]
+      - Library/include/*  # [win]
+      - Library/lib/*  # [win]
```
Comment (isuruf) [inline recipe/meta.yaml]: This would include unrelated `flang.lib`, right?

File: recipe/meta.yaml [hunk 8]
```diff
@@ -21,26 +20,59 @@ requirements:
...
+  - name: openblas
+    build:
+      skip: true  # [win and vc<14]
+    files:
+      - include/*.h  # [not win]
+      - lib/*.a  # [not win]
+      - lib/libopenblas${SHLIB_EXT}  # [not win]
+      - lib/pkgconfig/*.pc  # [not win]
+      - lib/cmake/openblas/*  # [not win]
+      - Library/include/*  # [win]
```
Comment (isuruf) [inline recipe/meta.yaml]: Same here

File: recipe/meta.yaml (hunk 5 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: OK. Then I'm not sure why we need to split it in a different package but since there is no harm let's go with it.

File: recipe/meta.yaml (hunk 8 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: You tell me  :smile:

File: recipe/meta.yaml (hunk 5 above)
Comment (isuruf) [reply recipe/meta.yaml]: > OK. Then I'm not sure why we need to split it in a different package but since there is no harm let's go with it.

To avoid linking to `libopenblas` at all. We want our downstream packages to link to `libblas.so` to allow switching the BLAS implementation

File: recipe/meta.yaml (hunk 7 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: The files in that folder are:

```
/Library/lib/pkgconfig/openblas.pc
openblas.lib
```

no `flang.lib`. I check the package `openblas-0.3.6-h828a276_2.tar.bz2`.

File: recipe/meta.yaml (hunk 8 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: The files in that folder are:

```
.
└── [4.0K]  openblas
    ├── [ 46K]  cblas.h
    ├── [ 47K]  f77blas.h
    ├── [4.3K]  lapacke_config.h
    ├── [ 359]  lapacke_example_aux.h
    ├── [1.2M]  lapacke.h
    ├── [ 474]  lapacke_mangling.h
    ├── [ 32K]  lapacke_utils.h
    ├── [4.0K]  openblas
    │   └── [ 474]  lapacke_mangling.h
    └── [5.0K]  openblas_config.h
```

File: recipe/meta.yaml (hunk 6 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: If you don't now I certainly don't :smile:

Comment (ocefpaf) [reply recipe/meta.yaml]: The files in there are:

```
.
└── [4.0K]  cmake
    └── [4.0K]  OpenBLAS
        ├── [3.6K]  OpenBLASConfig.cmake
        ├── [1.4K]  OpenBLASConfigVersion.cmake
        ├── [3.3K]  OpenBLASTargets.cmake
        └── [ 922]  OpenBLASTargets-release.cmake
```

File: recipe/meta.yaml (hunk 5 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: > To avoid linking to libopenblas at all. We want our downstream packages to link to libblas.so to allow switching the BLAS implementation

Then I'm back to my other question. Will we be able to switch `blas` if we accidentally use a `libopenblas` from `defaults` at build time?

Comment (isuruf) [reply recipe/meta.yaml]: No. You are right

Comment (ocefpaf) [reply recipe/meta.yaml]: I guess that we are still good to merge this as b/c `conda-forge`'s is newer and channel priority will help, but the only way to be really safe is to enable `strict` in our builds ASAP.

File: recipe/meta.yaml (hunk 6 above)
Comment (jakirkham) [reply recipe/meta.yaml]: These may be used by downstream packages leveraging CMake to build against OpenBLAS. We would want them in any kind of development package, which is what I understand this one to be. Though please correct me if I missed something.

**Comment by ocefpaf** [comment]
Thanks @isuruf for all the help and the discussion!

**Comment by jakirkham** [comment]
Thanks for tackling this one @ocefpaf. This is a nice win! 😄

**Comment by isuruf** [comment]
Needs a build number bump

**Comment by ocefpaf** [comment]
> Needs a build number bump

Yeah. I notice that as soon as I clicked the merge button.
